scio-core/src/main/scala/com/spotify/scio/ScioContext.scala

```diff
@@ -314,23 +312,37 @@ class ScioContext private[scio] (val options: PipelineOptions,
   private var _pipeline: Pipeline = _
   private var _isClosed: Boolean = false
   private val _promises: MBuffer[(Promise[Tap[_]], Tap[_])] = MBuffer.empty
-  private val _queryJobs: MBuffer[QueryJob] = MBuffer.empty
   private val _preRunFns: MBuffer[() => Unit] = MBuffer.empty
   private val _counters: MBuffer[Counter] = MBuffer.empty
+  private var _onClose: Unit => Unit = identity
+  private val _localInstancesCache: scala.collection.mutable.Map[ClassTag[_], Any] =
+    scala.collection.mutable.Map.empty
 
   /** Wrap a [[org.apache.beam.sdk.values.PCollection PCollection]]. */
   def wrap[T: ClassTag](p: PCollection[T]): SCollection[T] =
     new SCollectionImpl[T](p, this)
 
-  // =======================================================================
-  // Miscellaneous
-  // =======================================================================
+  /**
+  * Add callbacks calls when the context is closed.
+  */
+  private[scio] def onClose(f: Unit => Unit): Unit =
+    _onClose = _onClose compose f
 
-  private lazy val bigQueryClient: BigQueryClient = {
-    val o = optionsAs[GcpOptions]
-    BigQueryClient(o.getProject, o.getGcpCredential)
+  /**
+  * Get from or put in an object in this context local cache
+  * This method is used un scio-bigquery to only instanciate the bigquery client once
+  * even if there's multiple implicit conversions from [[ScioContext]] to [[BigQueryScioContext]]
+  */
+  private[scio] def cached[T: ClassTag](t: => T): T = {
```

Comment:
if you summon the key in next line you might as well desugar it to implicit argument?